packages/firestore/test/integration/api/database.test.ts

```diff
@@ -1340,6 +1368,125 @@ apiDescribe('Database', (persistence: boolean) => {
       });
     });
 
+    it('requires the correct converter for Partial usage', async () => {
+      return withTestDb(persistence, async db => {
+        const ref = db
+          .collection('posts')
+          .doc()
+          .withConverter(postConverter);
+        await ref.set(new Post('walnut', 'author'));
+        const batch = db.batch();
+        expect(() =>
+          batch.set(ref, { title: 'olive' }, { merge: true })
+        ).to.throw(
+          'Function toFirestore() in WriteBatch.set() called with invalid data. Unsupported field value: undefined (found in field author)'
+        );
```

Comment:
Ideally, this error message should prompt the user to provide a different converter.

Reply:
I'll try looking into this some more in a subsequent PR. There doesn't seem to be an easy way to differentiate between the two possible converters once its been compiled down. The only way I can think of is parsing `toFirestore.toString()` and checking if the options parameter is included in the function. Do you see an easy implementation here?

Reply:
The message is generated here: https://github.com/firebase/firebase-js-sdk/blob/f4451f05279ae534eb5f30a2ed579ca81062da3f/packages/firestore/src/api/database.ts#L2647

Isn't this message always wrong? The output of `toFirestore()` is invalid, not the input. Maybe change it to:

```
    functionName = functionName +  '() (converted by toFirestore()) ' +;
```

We need to change the way we pass the parenthesis though.

Since we are still waiting for sign off from @egilmorez, I vote we fix this in this PR.

Reply:
Done. Doing so cleans up `applyFirestoreDataConverter()` as well.

Reply:
Looks fantastic!